src/transformStateWithClones.js

```diff
@@ -1,13 +1,37 @@
 'use strict';
-
 /**
  * @param {Object} state
  * @param {Object[]} actions
  *
  * @return {Object[]}
  */
-function transformStateWithClones(state, actions) {
-  // write code here
+
+function transformStateWithClones(initialState, actions) {
+  const stateHistory = [];
+  let state = { ...initialState };
+
+  for (const action of actions) {
+    let newState;
+
+    if (action.type === 'clear') {
+      newState = {};
+    } else {
+      newState = { ...state };
+
+      if (action.type === 'addProperties') {
+        Object.assign(newState, action.extraData);
+      } else if (action.type === 'removeProperties') {
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+      }
```

Comment:
[CODE STYLE]: Use a switch statement instead of multiple if/else if conditions when you have a limited amount of conditions to check. This will make your code more readable and maintainable.